src/simd/distances_neon.cc

```diff
@@ -211,12 +316,99 @@ bf16_vec_inner_product_neon(const knowhere::bf16* x, const knowhere::bf16* y, si
 
 float
 fvec_L2sqr_neon(const float* x, const float* y, size_t d) {
-    float32x4_t sum_ = {0.0f, 0.0f, 0.0f, 0.0f};
+    float32x4_t sum_ = vdupq_n_f32(0.0f);
+    auto dim = d;
+    while (d >= 16) {
+        float32x4x4_t a = vld1q_f32_x4(x + dim - d);
+        float32x4x4_t b = vld1q_f32_x4(y + dim - d);
+        float32x4x4_t c;
+
+        c.val[0] = vsubq_f32(a.val[0], b.val[0]);
+        c.val[1] = vsubq_f32(a.val[1], b.val[1]);
+        c.val[2] = vsubq_f32(a.val[2], b.val[2]);
+        c.val[3] = vsubq_f32(a.val[3], b.val[3]);
+
+        c.val[0] = vmulq_f32(c.val[0], c.val[0]);
+        c.val[1] = vmulq_f32(c.val[1], c.val[1]);
+        c.val[2] = vmulq_f32(c.val[2], c.val[2]);
+        c.val[3] = vmulq_f32(c.val[3], c.val[3]);
+
+        c.val[0] = vaddq_f32(c.val[0], c.val[1]);
+        c.val[2] = vaddq_f32(c.val[2], c.val[3]);
+        c.val[0] = vaddq_f32(c.val[0], c.val[2]);
+
+        sum_ = vaddq_f32(sum_, c.val[0]);
+
+        d -= 16;
+    }
+
+    if (d >= 8) {
+        float32x4x2_t a = vld1q_f32_x2(x + dim - d);
+        float32x4x2_t b = vld1q_f32_x2(y + dim - d);
+        float32x4x2_t c;
+        c.val[0] = vsubq_f32(a.val[0], b.val[0]);
+        c.val[1] = vsubq_f32(a.val[1], b.val[1]);
+
+        c.val[0] = vmulq_f32(c.val[0], c.val[0]);
+        c.val[1] = vmulq_f32(c.val[1], c.val[1]);
+
+        c.val[0] = vaddq_f32(c.val[0], c.val[1]);
+        sum_ = vaddq_f32(sum_, c.val[0]);
+        d -= 8;
+    }
+    if (d >= 4) {
+        float32x4_t a = vld1q_f32(x + dim - d);
+        float32x4_t b = vld1q_f32(y + dim - d);
+        float32x4_t c;
+        c = vsubq_f32(a, b);
+        c = vmulq_f32(c, c);
+
+        sum_ = vaddq_f32(sum_, c);
+        d -= 4;
+    }
 
+    float32x4_t res_x = vdupq_n_f32(0.0f);
+    float32x4_t res_y = vdupq_n_f32(0.0f);
+    if (d >= 3) {
+        res_x = vld1q_lane_f32(x + dim - d, res_x, 2);
+        res_y = vld1q_lane_f32(y + dim - d, res_y, 2);
+        d -= 1;
+    }
+
+    if (d >= 2) {
+        res_x = vld1q_lane_f32(x + dim - d, res_x, 1);
+        res_y = vld1q_lane_f32(y + dim - d, res_y, 1);
+        d -= 1;
+    }
+
+    if (d >= 1) {
+        res_x = vld1q_lane_f32(x + dim - d, res_x, 0);
+        res_y = vld1q_lane_f32(y + dim - d, res_y, 0);
+        d -= 1;
+    }
+
+    sum_ = vaddq_f32(sum_, vmulq_f32(vsubq_f32(res_x, res_y), vsubq_f32(res_x, res_y)));
+
+    return vaddvq_f32(sum_);
+}
+
+float
+fvec_L2sqr_neon_bf16_patch(const float* x, const float* y, size_t d) {
+    float32x4_t sum_ = vdupq_n_f32(0.0f);
```

Comment:
same here and in the following
Basically, please refer to `distances_avx512.cpp` to see which functions are delegated to a compiler to be SIMD-ed properly